python/hsfs/core/opensearch_api.py

```diff
@@ -0,0 +1,136 @@
+#
+#   Copyright 2023 Logical Clocks AB
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+
+import os
+from furl import furl
+
+from hsfs import client
+from hsfs.client.exceptions import FeatureStoreException
+from hsfs.core.variable_api import VariableApi
+
+
+class OPENSEARCH_CONFIG:
+    ELASTIC_ENDPOINT_ENV_VAR = "ELASTIC_ENDPOINT"
+    SSL_CONFIG = "es.net.ssl"
+    NODES_WAN_ONLY = "es.nodes.wan.only"
+    NODES = "es.nodes"
+    SSL_KEYSTORE_LOCATION = "es.net.ssl.keystore.location"
+    SSL_KEYSTORE_PASSWORD = "es.net.ssl.keystore.pass"
+    SSL_TRUSTSTORE_LOCATION = "es.net.ssl.truststore.location"
+    SSL_TRUSTSTORE_PASSWORD = "es.net.ssl.truststore.pass"
+    HTTP_AUTHORIZATION = "es.net.http.header.Authorization"
+    INDEX = "es.resource"
+    HOSTS = "hosts"
+    HTTP_COMPRESS = "http_compress"
+    HEADERS = "headers"
+    USE_SSL = "use_ssl"
+    VERIFY_CERTS = "verify_certs"
+    SSL_ASSERT_HOSTNAME = "ssl_assert_hostname"
+    CA_CERTS = "ca_certs"
+
+
+class OpenSearchApi:
+    def __init__(
+        self,
+        project_id,
+        project_name,
+    ):
+        self._project_id = project_id
+        self._project_name = project_name
+        self._variable_api = VariableApi()
+
+    def _get_opensearch_url(self):
+        url = os.environ.get(OPENSEARCH_CONFIG.ELASTIC_ENDPOINT_ENV_VAR)
```

Comment:
We should not be dependent on this ELASTIC_ENDPOINT environment variable. There may be environments where this is not present and that will break the API. Since we have service discovery (consul) for applications within our private network it should use that.

My understanding is we should use the same logic for determining the opensearch url as we do for flying duck:
https://github.com/logicalclocks/feature-store-api/blob/6ed179af26d1005b72b971b88b28cd4724333646/python/hsfs/core/arrow_flight_client.py#L87

That being said I'm not an expert in this domain but the current logic is not correct. On line 77 you do `return f"https://{service_discovery_domain}:9200"` which effectively just becomes `https://consul:9200`?

I'm happy for maybe @SirOibaf which reviewed this corresponding code for flying duck to chime in here.

Reply:
We only use it if it is available: https://github.com/kennethmhc/feature-store-api/blob/FSTORE-1047/python/hsfs/core/opensearch_api.py#L57 otherwise we infer the url end point using the flying duck logic: https://github.com/kennethmhc/feature-store-api/blob/FSTORE-1047/python/hsfs/core/opensearch_api.py#L60 (mostly copy from https://github.com/logicalclocks/feature-store-api/blob/master/python/hsfs/core/arrow_flight_client.py#L87)

Reply:
Yes I understand that but the code is more clean and consistent if we have the same logic for determining the endpoint for different services. Checking that environment variable is added complexity in my view and for anyone that didn't write the code.

I saw it was a copy, but for example this line, https://github.com/logicalclocks/feature-store-api/blob/60bfb1a74811b323ed014afa40fc9468e5d23053/python/hsfs/core/arrow_flight_client.py#L102, as I mentioned before I still don't understand how that can work? From my perspective it should be something like:

`host_url = f"https://elastic.service.{service_discovery_domain}:9200"`

Reply:
right it should be `f"https://elastic.service.{service_discovery_domain}:9200"`. I think it is necessary to pass the endpoint from environment variable as well, as it should also work with external elastic cluster whose url is different from the hopsworks backend.

Reply:
The environment variable is not necessary, if you are external to the cluster, then we need to setup the load balancer, if you are inside the cluster, we can use the service discovery and it doesn't matter where elastic is running, it will give back the right IP.

Additionally using the service discovery allows us to handle cases where one elastic node goes down. the client will do the resolution again and consul will give back another IP.

Btw the full name of the elastic dns is `rest.elastic.service.`

Reply:
@robzor92 @SirOibaf Removed the environment variable and change the elastic dns to `rest.elastic.service`
https://github.com/logicalclocks/feature-store-api/pull/1139/commits/85300703999f2f567700f57efadf0d25f73d3309